fbgemm_gpu/codegen/embedding_backward_split_meta_template.cpp

```diff
@@ -0,0 +1,218 @@
+{%- endif %}
+    const Tensor& grad_output,
+    const Tensor& dev_weights,
+    {%- if not dense %}
+    const Tensor& uvm_weights,
+    const Tensor& lxu_cache_weights,
+    const Tensor& weights_placements,
+    {%- endif %}
+    const Tensor& weights_offsets,
+    {%- if not nobag or is_index_select %}
+    const Tensor& D_offsets,
+    const int64_t max_D,
+    {%- else %}
+    const int64_t D,
+    {%- endif %}
+    const Tensor& hash_size_cumsum,
+    const int64_t total_hash_size_bits,
+    const Tensor& indices,
+    {%- if not is_index_select %}
+    const Tensor& offsets,
+    {%- endif %}
+    {%- if not nobag %}
+    const int64_t pooling_mode,
+    {%- endif %}
+    {%- if weighted %}
+    const Tensor& indice_weights,
+    {%- endif %}
+    {%- if not dense %}
+    const Tensor& lxu_cache_locations,
+    {%- endif %}
+    {%- if not is_index_select %}
+    const int64_t unused_,
+    {%- endif %}
+    const int64_t max_segment_length_per_warp,
+    {%- if not dense %}
+    {%- if optimizer != "none" %}
+    const bool stochastic_rounding,
+    {%- endif %}
+    const int64_t info_B_num_bits, // int32_t
+    const int64_t info_B_mask_int64, // uint32_t
+    {%- endif %}
+    {%- if vbe %}
+    const Tensor& B_offsets,
+    const Tensor& vbe_row_output_offsets,
+    const Tensor& vbe_b_t_map,
+    {%- endif %}
+    {%- if not is_index_select and not dense %}
+    const bool use_uniq_cache_locations,
+    const bool use_homogeneous_placements,
+    {%- endif %}
+    {%- if is_index_select %}
+    const Tensor& grad_offsets,
+    const Tensor& total_L_offsets,
+    const int32_t fixed_L_per_warp,
+    const int32_t num_warps_per_feature,
+    const bool permute_output_dim_0_1
+    {%- elif optimizer != "none" %}
+    {{ args.split_function_args_no_defaults | join(", ") }}
+    {%- else %}
+    // This is actually passed via args.split_function_args_no_defaults but explicitly list
+    // it here for code readability
+    int64_t total_hash_size,
+    int64_t total_unique_indices
+    {%- endif %}
+) {
+
+    // NB: Should we have something for aligning memory like we do on the tensor kernels?
+    // There it checks if the pointer to the tensor is not divisble by 16, along with
+    // some stride checks.
+
+    {%- if nobag and not is_index_select %}
+    auto max_D = D;
+    {%- endif %}
+    {%- if not is_index_select %}
+    TORCH_CHECK_LE(max_D, {{ max_embedding_dim }});
+    {%- endif %}
+
+    {%- if optimizer == "none" %}
+    // grad_dev_weights has emb_t type
+    auto grad_dev_weights = at::empty_symint({total_unique_indices * max_D}, dev_weights.options());
+    {%- else %}
+    // Set total_unique_indices to total num indices by default
+    const auto total_unique_indices = indices.sym_numel();
+    {%- if dense %}
+    auto grad_dev_weights = at::zeros_like(dev_weights);
+    {%- endif %}
+    {%- endif %}
+
+    // short-circuit if there are zero indices.
+    if (indices.sym_numel() == 0) {
```

Comment:
Make this one size oblivious, per https://docs.google.com/document/d/1HSuTTVvYH1pTew89Rtpeu84Ht3nQEFTYhAX3Ypa_xJs/edit#heading=h.11jnmcqhq5yy